engines/CaesarEngine.py

```diff
@@ -0,0 +1,66 @@
+from cipher.cipher import Cipher
+from cipher.alphabet import Alphabet
+
+
+class CaesarCipher(Cipher):
+
+    def __init__(self, spec):
+        key, plain_range, cipher_range = spec.split("/")
+        self.alphabet = Alphabet(plain_range, cipher_range)
+        try:
+            self.key = int(key)
+        except ValueError:
+            raise Exception("Key should be int")
+        if len(self.alphabet.plain_alphabet) != len(
+                self.alphabet.cipher_alphabet):
+            raise Exception("Length of both alphabet arrays not equal")
+
+    def encode_shift(self, ch, key):
+        plain_alphabet = self.alphabet.plain_alphabet
+        cipher_alphabet = self.alphabet.cipher_alphabet
+        if ord(ch) not in plain_alphabet:
+            return ord(ch)
+        while key < 0:
```

Comment:
You are branching in each call, expect that key is positive and move the check to the caller. You can use a `__debug__` (https://docs.python.org/3/library/constants.html#__debug__) to check and complain in debug mode